onecodex/helpers.py

```diff
@@ -1,169 +1,328 @@
+                    magic_metadata[renamed_field] = df[str_f]
+                    magic_fields[f] = renamed_field
+                else:
+                    # try to match it up with a taxon name
+                    hits = []
+
+                    # don't both searching if the query is really short
+                    if len(str_f) > 4:
+                        for tax_id, tax_name in zip(self.taxonomy.index, self.taxonomy['name']):
+                            # if it's an exact match, use that and skip the rest
+                            if str_f.lower() == tax_name.lower():
+                                hits = [(tax_id, tax_name)]
+                                break
+                            # otherwise, keep trying to match
+                            elif str_f.lower() in tax_name.lower():
+                                hits.append((tax_id, tax_name))
+
+                        # take the hit with the lowest tax_id
+                        hits = sorted(hits, key=lambda x: int(x[0]))
+
+                    if hits:
+                        # report within-rank abundance
+                        df = self.results(rank=self.taxonomy['rank'][hits[0][0]])
+
+                        renamed_field = '{} ({})'.format(hits[0][1], hits[0][0])
+                        magic_metadata[renamed_field] = df[hits[0][0]]
+                        magic_fields[f] = renamed_field
+                    else:
+                        # matched nothing
+                        raise OneCodexException(
+                            'Field or taxon {} not found. Choose from: {}'.format(str_f, help_metadata)
+                        )
+
+        return magic_metadata, magic_fields
+
+    def results(self, rank='auto', top_n=None, threshold=None,
+                remove_zeros=True, normalize='auto',
+                table_format='wide'):
+        """Takes the ResultsDataFrame associated with these samples, or SampleCollection, does some
+        filtering, and returns a ResultsDataFrame.
+
+        Parameters
+        ----------
+        rank : {'auto', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'}, optional
+            Analysis will be restricted to abundances of taxa at the specified level.
+        top_n : `integer`, optional
+            Return only the top N most abundant taxa.
+        threshold : `float`, optional
+            Return only taxa more abundant than this threshold in one or more samples.
+        remove_zeros : `bool`, optional
+            Do not return taxa that have zero abundance in every sample.
+        normalize : {'auto', True, False}
+            Convert read counts to relative abundances (each sample sums to 1.0).
+        table_format : {'long', 'wide'}
+            If wide, rows are classifications, cols are taxa, elements are counts. If long, rows are
+            observations with three cols each: classification_id, tax_id, and count.
```

Comment:
Would it be possible to split up all of these transformation steps into separate public functions? It would be easier to test and I think it's clearer (although the style you've chosen _is_ more similar to Pandas' conventions).